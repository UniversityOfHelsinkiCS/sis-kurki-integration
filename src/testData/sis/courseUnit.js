export default {
  id: 'hy-CU-118901507-2019-08-01',
  groupId: 'hy-CU-118901507',
  code: 'TKT21002',
  name: {
    en: 'Introduction to Game Programming',
    fi: 'Introduction to Game Programming',
    sv: 'Introduction to Game Programming',
  },
  validityPeriod: { endDate: '2020-08-01', startDate: '2019-08-01' },
  gradeScaleId: 'sis-0-5',
  studyLevel: 'urn:code:study-level:intermediate-studies',
  courseUnitType: 'urn:code:course-unit-type:regular',
  possibleAttainmentLanguages: ['urn:code:language:en'],
  assessmentItemOrder: null,
  organisations: [
    {
      share: 1,
      roleUrn: 'urn:code:organisation-role:responsible-organisation',
      organisationId: 'hy-org-116716376',
      validityPeriod: {},
      educationalInstitutionUrn: null,
    },
  ],
  universityOrgIds: ['hy-university-root-id'],
  studyFields: ['urn:code:study-field:okm-7'],
  substitutions: [],
  completionMethods: [
    {
      localId: 'hy-CM-118901507-5-5-106',
      repeats: [],
      require: null,
      description: null,
      typeOfRequire: 'ALL_SELECTED_REQUIRED',
      assessmentItemIds: ['hy-AI-118901507-5-5-106'],
      evaluationCriteria: null,
      automaticEvaluation: true,
      assessmentItemOptionalityDescription: null,
    },
    {
      localId: 'hy-CM-118901507-8-5-106',
      repeats: [],
      require: null,
      description: null,
      typeOfRequire: 'ALL_SELECTED_REQUIRED',
      assessmentItemIds: ['hy-AI-118901507-8-5-106'],
      evaluationCriteria: null,
      automaticEvaluation: true,
      assessmentItemOptionalityDescription: null,
    },
  ],
  responsibilityInfos: [
    {
      text: null,
      roleUrn: 'urn:code:module-responsibility-info-type:responsible-teacher',
      personId: 'hy-hlo-1486440',
      validityPeriod: {},
    },
  ],
  createdAt: '2020-02-04T13:18:17.488Z',
  updatedAt: '2020-03-17T10:34:54.339Z',
};
