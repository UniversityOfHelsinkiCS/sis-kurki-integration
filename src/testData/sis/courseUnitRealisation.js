export default {
  id: 'hy-CUR-119284687',
  universityOrgIds: ['hy-university-root-id'],
  flowState: 'PUBLISHED',
  name: { fi: 'Introduction to Game Programming' },
  nameSpecifier: { fi: 'Luentokurssi' },
  assessmentItemIds: ['hy-AI-118901507-5-5-106'],
  activityPeriod: { endDate: '2017-12-16', startDate: '2017-11-01' },
  teachingLanguageUrn: 'urn:code:language:en',
  courseUnitRealisationTypeUrn:
    'urn:code:course-unit-realisation-type:teaching-participation-lectures',
  studyGroupSets: [
    {
      name: { fi: 'Luentokurssi' },
      localId: 'hy-SGS-119284687',
      subGroupRange: { max: null, min: 0 },
      studySubGroups: [
        {
          id: 'hy-SSG-119284687',
          name: { fi: 'Introduction to Game Programming' },
          size: null,
          metadata: null,
          cancelled: false,
          teacherIds: ['hy-hlo-1486440'],
          studyEventIds: [
            'hy-event-121960900',
            'hy-event-121960899',
            'hy-event-121960898',
            'hy-event-121960776',
            'hy-event-121524681',
            'hy-event-121524713',
          ],
        },
      ],
    },
    {
      name: { fi: 'Harjoitusryhmä' },
      localId: 'hy-SGS-119284691',
      subGroupRange: { max: null, min: 1 },
      studySubGroups: [
        {
          id: 'hy-SSG-119284691',
          name: { fi: 'Ryhmä 1' },
          size: null,
          metadata: null,
          cancelled: false,
          teacherIds: [],
          studyEventIds: ['hy-event-121524715'],
        },
        {
          id: 'hy-SSG-119284693',
          name: { fi: 'Ryhmä 2' },
          size: null,
          metadata: null,
          cancelled: false,
          teacherIds: [],
          studyEventIds: ['hy-event-121524716'],
        },
        {
          id: 'hy-SSG-121532097',
          name: {
            fi: 'Ryhmä 99 (Jono - jos ryhmät ovat täynnä tai ajat eivät sovi)',
          },
          size: null,
          metadata: null,
          cancelled: false,
          teacherIds: [],
          studyEventIds: [],
        },
      ],
    },
  ],
  organisations: [
    {
      share: 1,
      roleUrn: 'urn:code:organisation-role:responsible-organisation',
      organisationId: 'hy-org-116716376',
      validityPeriod: {},
      educationalInstitutionUrn: null,
    },
  ],
  enrolmentPeriod: null,
  createdAt: '2020-02-04T13:48:41.706Z',
  updatedAt: '2020-03-17T10:59:31.611Z',
};
