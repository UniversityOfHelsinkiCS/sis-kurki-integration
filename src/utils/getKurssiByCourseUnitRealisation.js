import { get } from 'lodash';
import { getYear, isValid, getMonth, getDate } from 'date-fns';

const kielikoodiByLanguageCode = {
  fi: 'S',
  en: 'E',
  se: 'R',
};

const tyyppiByCourseUnitRealisationType = {
  'independent-work-essay': 'K',
  'exam-electronic': 'L',
  'exam-final': 'L',
  'teaching-participation-field-course': 'K',
  'teaching-participation-small-group': 'K',
  'independent-work-project': 'A',
  'teaching-participation-seminar': 'S',
  'thesis-doctoral': 'Y',
  'licentiate-thesis': 'Y',
  'independent-work-presentation': 'K',
  'training-training': 'K',
  'exam-exam': 'L',
  'teaching-participation-lab': 'A',
  'exam-midterm': 'L',
  'independent-work-learning-diary': 'K',
  'teaching-participation-lectures': 'K',
  'teaching-participation-online': 'K',
};

const tilaByFlowState = {
  PUBLISHED: 'I',
  CANCELLED: 'J',
  NOT_READY: 'O',
};

const getLukuvuosi = (dateLike) => {
  const date = new Date(dateLike);

  if (!isValid(date)) {
    return undefined;
  }

  return getYear(date);
};

const getLukukausi = (dateLike) => {
  const date = new Date(dateLike);

  if (!isValid(date)) {
    return undefined;
  }

  // Month index starts at 0
  const month = getMonth(date) + 1;
  const dayOfMonth = getDate(date);

  if (month < 5) {
    return 'K';
  } else if (month > 8 || (month === 8 && dayOfMonth > 20)) {
    return 'S';
  }

  return 'V';
};

const getKielikoodiByTeachingLanguageUrn = (teachingLanguageUrn) => {
  if (!teachingLanguageUrn) {
    return undefined;
  }

  const parts = teachingLanguageUrn.split(':');
  const language = parts[parts.length - 1];

  return kielikoodiByLanguageCode[language] || 'E';
};

const getTyyppiByCourseUnitRealisationTypeUrn = (
  courseUnitRealisationTypeUrn,
) => {
  if (!courseUnitRealisationTypeUrn) {
    return undefined;
  }

  const parts = courseUnitRealisationTypeUrn.split(':');
  const type = parts[parts.length - 1];

  return tyyppiByCourseUnitRealisationType[type] || 'K';
};

const getTilaByFlowState = (flowState) => {
  if (!flowState) {
    return undefined;
  }

  return tilaByFlowState[flowState] || 'O';
};

const getKurssiByCourseUnitRealisation = (
  courseUnitRealisation,
  courseUnit,
) => {
  const {
    activityPeriod,
    teachingLanguageUrn,
    courseUnitRealisationTypeUrn,
    flowState,
    name,
  } = courseUnitRealisation;

  const { code } = courseUnit;
  const startDate = get(activityPeriod, 'startDate');

  return {
    kurssikoodi: code,
    lukuvuosi: getLukuvuosi(startDate),
    lukukausi: getLukukausi(startDate),
    tyyppi: getTyyppiByCourseUnitRealisationTypeUrn(
      courseUnitRealisationTypeUrn,
    ),
    kurssiNro: 1,
    kielikoodi: getKielikoodiByTeachingLanguageUrn(teachingLanguageUrn),
    opintoviikot: 1,
    nimi: get(name, 'fi'),
    tila: getTilaByFlowState(flowState),
  };
};

export default getKurssiByCourseUnitRealisation;
