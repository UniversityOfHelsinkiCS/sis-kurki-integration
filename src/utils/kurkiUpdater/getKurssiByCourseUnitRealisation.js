import { get } from 'lodash';
import { getYear, isValid, getMonth, getDate } from 'date-fns';

import getLastUrnPart from '../getLastUrnPart';

const kielikoodiByLanguageCode = {
  fi: 'S',
  en: 'E',
  se: 'R',
};

const tyyppiByCourseUnitRealisationType = {
  'independent-work-essay': 'K',
  'exam-electronic': 'L',
  'exam-final': 'L',
  'teaching-participation-field-course': 'K',
  'teaching-participation-small-group': 'K',
  'independent-work-project': 'A',
  'teaching-participation-seminar': 'S',
  'thesis-doctoral': 'Y',
  'licentiate-thesis': 'Y',
  'independent-work-presentation': 'K',
  'training-training': 'K',
  'exam-exam': 'L',
  'teaching-participation-lab': 'A',
  'exam-midterm': 'L',
  'independent-work-learning-diary': 'K',
  'teaching-participation-lectures': 'K',
  'teaching-participation-online': 'K',
};

const tilaByFlowState = {
  PUBLISHED: 'I',
  CANCELLED: 'J',
  NOT_READY: 'O',
};

const getLukuvuosi = (dateLike) => {
  const date = new Date(dateLike);

  if (!isValid(date)) {
    return undefined;
  }

  return getYear(date);
};

const getLukukausi = (dateLike) => {
  const date = new Date(dateLike);

  if (!isValid(date)) {
    return undefined;
  }

  // Month index starts at 0
  const month = getMonth(date) + 1;
  const dayOfMonth = getDate(date);

  if (month < 5) {
    return 'K';
  } else if (month > 8 || (month === 8 && dayOfMonth > 20)) {
    return 'S';
  }

  return 'V';
};

const getKielikoodiByTeachingLanguageUrn = (teachingLanguageUrn) => {
  if (!teachingLanguageUrn) {
    return undefined;
  }

  const language = getLastUrnPart(teachingLanguageUrn);

  return kielikoodiByLanguageCode[language] || 'E';
};

const getTyyppiByCourseUnitRealisationTypeUrn = (
  courseUnitRealisationTypeUrn,
) => {
  if (!courseUnitRealisationTypeUrn) {
    return undefined;
  }

  const type = getLastUrnPart(courseUnitRealisationTypeUrn);

  return tyyppiByCourseUnitRealisationType[type] || 'K';
};

const getTilaByFlowState = (flowState) => {
  if (!flowState) {
    return undefined;
  }

  return tilaByFlowState[flowState] || 'O';
};

const getKurssiByCourseUnitRealisation = (courseUnitRealisation) => {
  const {
    activityPeriod,
    teachingLanguageUrn,
    courseUnitRealisationTypeUrn,
    flowState,
    name,
  } = courseUnitRealisation;

  const startDate = get(activityPeriod, 'startDate');
  const endDate = get(activityPeriod, 'endDate');

  return {
    sisId: courseUnitRealisation.id,
    lukuvuosi: getLukuvuosi(startDate),
    lukukausi: getLukukausi(startDate),
    tyyppi: getTyyppiByCourseUnitRealisationTypeUrn(
      courseUnitRealisationTypeUrn,
    ),
    kielikoodi: getKielikoodiByTeachingLanguageUrn(teachingLanguageUrn),
    opintoviikot: 1,
    nimi: get(name, 'fi'),
    tila: getTilaByFlowState(flowState),
    alkamisPvm: startDate ? new Date(startDate) : null,
    paattymisPvm: endDate ? new Date(endDate) : null,
  };
};

export default getKurssiByCourseUnitRealisation;
