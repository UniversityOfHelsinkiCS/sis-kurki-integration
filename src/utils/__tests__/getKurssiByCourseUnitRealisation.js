import courseUnit from '../../testData/sis/courseUnit';
import courseUnitRealisation from '../../testData/sis/courseUnitRealisation';
import getKurssiByCourseUnitRealisation from '../getKurssiByCourseUnitRealisation';

describe('getKurssiByCourseUnitRealisation', () => {
  it('returns correct kurssi given course unit realisation', () => {
    expect(
      getKurssiByCourseUnitRealisation(courseUnitRealisation, courseUnit),
    ).toMatchSnapshot();
  });
});
