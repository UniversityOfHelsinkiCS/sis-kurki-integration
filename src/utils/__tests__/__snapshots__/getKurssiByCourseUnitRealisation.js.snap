// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getKurssiByCourseUnitRealisation returns correct kurssi given course unit realisation 1`] = `
Object {
  "kielikoodi": "E",
  "kurssiNro": 1,
  "kurssikoodi": "TKT21002",
  "lukukausi": "S",
  "lukuvuosi": 2017,
  "nimi": "Introduction to Game Programming",
  "opintoviikot": 1,
  "tila": "I",
  "tyyppi": "K",
}
`;
